platform-plugins.md

```diff
@@ -0,0 +1,277 @@
+ String clipboardContents = 'Unknown';
+
+ getClipboardContents() async {
+   var clipboardData = await Clipboard.getData(Clipboard.kTextPlain);
+   if (clipboardData != null) {
+     setState(() {
+       clipboardContents = clipboardData.text;
+     });
+   }
+ }
+ ...
+}
+```
+
+### Platform Plugins from the Flutter ecosystem
+
+{% include note.html content="Sharing of Platform Plugins is currently not
+supported, but Flutter is expected to support this in the Future." %}
+
+## Create a Platform Plugin {#create}
+
+Flutter uses a flexible system that allows you to call *any* platform-specific
+API whether available in Java-code on Android, or in ObjectiveC- or Swift-code
+on iOS.
+
+Flutter's platform-specific API support does not rely on fragile code
+generation, but rather on a flexible message passing style:
+
+* The Flutter portion of your app sends messages to it's *host*, the iOS or
+ Android portion of your app, over a platform channel.
+
+* The *host* listens on the platform channel, and receives the message. It then
+ calls into any number of platform-specific APIs -- using their native
+ programming language -- and sends back a response to the *client*, the Flutter
+ portion.
+
+### Architectural overview: Platform Channels
+
+Messages are passed between the client (UI) and host (platform) using Platform
+channels as illustrated in this diagram:
+
+![Platform Channels architecture](/images/PlatformChannels.png)
+
+Message and responses are passed asynchronously, to ensure the user interface
+remains responsive.
+
+On the client side, the class
+[`PlatformMessageChannel`](https://docs.flutter.io/flutter/services/PlatformMessageChannel-class.html)
+enables sending general purpose message, and the class
+[`PlatformMethodChannel`](https://docs.flutter.io/flutter/services/PlatformMethodChannel-class.html)
+enables spending specific messages that correspond to method calls. On the
+platform side, the class `FlutterMessageChannel` enables receiving messages
+messages from the client, and responding with a result, and the class
+`FlutterMethodChannel`. These classes allow you to developer a platform plugin
+with very little 'boilerplate' code.
+
+### Example: Calling platform-specific iOS and Android code using Platform Channels
+
+The following example demonstrates how to call a new platform-specific API,
+through a concrete example that retrieves and displays the current battery level
+by calling into the Android `BatteryManager` API, and the iOS
+`device.batteryLevel` API, via a single platform message, `getBatteryLevel`.
+
+The standard app template created when you create a new Flutter app can be used
+
+*Note*: The full source-code for this example is available in
+[`/samples/platform-services/`](https://github.com/flutter/flutter/tree/master/examples/platform_services).
+
+**Step 1: Create the Flutter platform client**
+
+The Flutter platform client code is written inside the app's `State` class.
```

Comment:
This read as if we always do this. Do we want every user of PlatformMethodChannel to be a static const inside a State class?

Also, the app has many State classes.

Maybe: "it's common practice to create a State class that manages a PlatformMethodChannel" or similar?